plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,443 @@
+    index_h_vec.push_back(index_h);
+  }
+  for (auto off_w = 0; off_w < OFF_W - 1; ++off_w) {
+    index_w = index_w + strides[2];
+    index_w_vec.push_back(index_w);
+  }
+  index_d = op::concatenate(index_d_vec, 0);
+  index_h = op::concatenate(index_h_vec, 0);
+  index_w = op::concatenate(index_w_vec, 0);
+  index_d = index_d - pad_befores[0];
+  index_h = index_h - pad_befores[1];
+  index_w = index_w - pad_befores[2];
+  index_d = op::reshape(index_d, make_tuple<int64_t>({1, 1, NEW_D, 1, 1}));
+  index_h = op::reshape(index_h, make_tuple<int64_t>({1, 1, 1, NEW_H, 1}));
+  index_w = op::reshape(index_w, make_tuple<int64_t>({1, 1, 1, 1, NEW_W}));
+  index_d = op::broadcast(index_d, {1, 1, static_cast<int>(NEW_D), static_cast<int>(NEW_H), static_cast<int>(NEW_W)},
+                          {0, 1, 2, 3, 4});
+  index_h = op::broadcast(index_h, {1, 1, static_cast<int>(NEW_D), static_cast<int>(NEW_H), static_cast<int>(NEW_W)},
+                          {0, 1, 2, 3, 4});
+  index_w = op::broadcast(index_w, {1, 1, static_cast<int>(NEW_D), static_cast<int>(NEW_H), static_cast<int>(NEW_W)},
+                          {0, 1, 2, 3, 4});
+  // Get deformabled index.
+  edsl::Tensor new_index_d = offset_d + index_d;
+  edsl::Tensor new_index_h = offset_h + index_h;
+  edsl::Tensor new_index_w = offset_w + index_w;
+  // Get deformable input tensor.
+  edsl::Tensor deform_input = edsl::gather(I, new_index_d).axis(-3);
+  deform_input = edsl::gather(deform_input, new_index_h).axis(-2);
+  deform_input = edsl::gather(deform_input, new_index_w).axis(-1);
+  deform_input = extract_tensor(deform_input, rank - 2);
+  // Compute DeformableConvolution.
+  autopad_mode = plaidml::op::AutoPadMode::VALID;
+  auto result = op::convolution(deform_input, F)
+                    .strides(std::vector<size_t>({(size_t)F_D, (size_t)F_H, (size_t)F_W}))
+                    .autopad_mode(autopad_mode)
+                    .input_layout(plaidml::op::TensorLayout::NCX)
+                    .filter_layout(plaidml::op::TensorLayout::KCX)
+                    .autogroup_mode(plaidml::op::AutoGroupMode::EXPLICIT)
+                    .group_layout(plaidml::op::GroupLayout::IN_C)
+                    .group_layout(plaidml::op::GroupLayout::IN_K)
+                    .groups(static_cast<int>(G));
+  return result;
+}
+
+void registerDeformableConvolution() {
+  registerOp("DeformableConvolution", [](const Context& ctx) {
+    auto* layer = ngraph::as_type<ngraph::opset4::DeformableConvolution>(ctx.layer);
+    IE_ASSERT(ctx.operands.size() == 3);
+    auto I = ctx.operands.at(0), OFF = ctx.operands.at(1), F = ctx.operands.at(2);
+    auto I_shape = I.compute_shape().sizes(), OFF_shape = OFF.compute_shape().sizes(),
+         F_shape = F.compute_shape().sizes();
+    auto G = layer->get_group(), DG = layer->get_deformable_group();
+    auto rank = I.rank();
+    // Get autopad_mode.
+    auto autopad_mode = to_plaidml(layer->get_auto_pad());
+    // Compute manual_padding.
```

Comment:
This section can be clearer by using the following:
```
 std::vector<size_t> manual_padding;
 if (autopad_mode == plaidml::op::AutoPadMode::EXPLICIT) {
    auto pads_begin=layer->get_pads_begin();
    auto pads_end=layer->get_pads_end();
    manual_padding.insert(manual_padding.end(),pads_begin.begin(),pads_begin.end());
    manual_padding.insert(manual_padding.end(),pads_end.begin(),pads_end.end());
 }
 while (manual_padding.size() < 2 * (rank - 2)) {
    manual_padding.push_back(0);
 }
```

Reply:
A question here: is the pad vector size of `layer->get_pads_begin()` always equals rank?

Reply:
There is no detail in openvino, I wirte it like the function of convolution in plaidml/op/lib/ops.cc